### ISSUE TITLE
Use seaborn model of dataset loading

### ISSUE THREAD

**Issue by utf** [hackingmaterials/matminer#307]
Currently datasets are loaded by calling a specific function. E.g.:

```python
from matminer.datasets.dataframe_loader import load_elastic_tensor
df = load_elastic_tensor()
```

If we want to have multiple elastic datasets (e.g. one generated from Materials Project data and one from Citrine), this will quickly become messy. Instead, it would be nice to use the Seaborn model of loading datasets. E.g.:

```python
import seaborn as sns
iris = sns.load_dataset('iris')
```

In our case, we can use specific strings to identify datasets. E.g. `'elastic_tensor_Ward_2015'`, vs `'elastic_tensor_Jain_2018'`. We should make it clear what datasets are available both in the documentation and via the API (e.g. a `matminer.datasets.available_datasets` list).

Currently, the implementation in matminer performs dataset specific pre-processing to convert from a csv file to the Pandas dataframe. Instead, we should use the json serialization features to store the processed datasets, which can then be loaded without need for further adjustment. We should also use the compression features to minimize the dataset size.

It makes sense that this feature be implemented after #297. After chatting with @Doppe1g4nger, we think the easiest way to implement this would be a `matminer.datasets` module level dictionary, containing something like:

```python
datasets_data = {
	'elastic_tensor_Ward_2018': {
		'url': 'url_of_dataset',
		'citation': 'bibliography_data'
		'metadata': ['list of optional database columns']
	}
}
```

**Comment by ardunn** [comment]
I agree with this! Converting between csv and inflated dataframe each time a dataset is loaded is not ideal.

**Comment by computron** [comment]
Some things I like about CSV

- you can include a comment at the top to give file provenance (sometimes nice to have this in the file itself)
- people can load it easily into other software they have (Excel, various other statistics software, etc)

That said, neither of these are particularly strong reasons to stick with CSV. So I'd say go with whatever you guys decide

**Comment by computron** [comment]
Also the string data set loader sounds good to me (better than individual functions for each data set)

**Comment by Doppe1g4nger** [comment]
I think it would make sense to implement this in two stages. First with the generic load_dataset function using the current csv files, then later we can consider switching over to JSON. 

My only concern with going over to this method is that I would like to make sure it is easily importable to other projects. When making #297 I tried to write the code in such a way that you can import some of the underlying functions to easily write additional load_x style functions in other packages (I had matbench in particular in mind). If we use a module level dictionary of available datasets then the use of the function becomes strongly tied to matminer.

I think that can be fixed by writing it as a DatasetLoader class instead of a single function. The class can have a default set of datasets specified as a dictionary, a class method for displaying available datasets, and a method that loads each dataset based on string keyword. This class could then be imported and subclassed in other projects with the project specific datasets overriding the default ones.

**Comment by utf** [comment]
I agree about implementing this in two parts. Although it does seem a little unwieldy to implement this as its own class (especially when the aim is to have this as neat as possible). 

Can you just make a function called `load_dataset_from_dict()` which expects a dictionary with the url, and metadata keys described above. This function is the importable into any project you want.

Matminer then implements its own method called `load_dataset()` defined as:

```python
from matminer.datasets import dataset_dict

def load_dataset(dataset):
    return load_dataset_from_dict(dataset_dict[dataset])
```

It's then not much work to add a method in matbench as:

```python
from matminer.datasets import load_dataset_from_dict
from matbench.datasets import dataset_dict

def load_dataset(dataset):
    return load_dataset_from_dict(dataset_dict[dataset])
```

**Comment by utf** [comment]
Concentrating the dataset loading in matminer would make sense from a project scope point of view. E.g.
- Matminer is used for loading data, featurizing, and plotting.
- Matbench is for quickly & easily obtaining a high performance model.

In that case the file storage location would be controlled by matminer. So either a custom folder if the correct environment variable is set, or the `matminer.datasets` folder if not.

**Comment by ardunn** [comment]
I vote for @utf idea

The datasets should be accessible from other projects (e.g., matbench) via the environment variable. So I don't see the 3rd "con" @Doppe1g4nger mentioned being particularly problematic.

In other words, all pros!
